Documentation/unicorn.1
.TH "UNICORN" "1" "September 15, 2009" "Unicorn User Manual" ""
.hy
.SH NAME
.PP
unicorn \- a rackup\-like command to launch the Unicorn HTTP server
.SH SYNOPSIS
.PP
unicorn [\-c CONFIG_FILE] [\-E RACK_ENV] [\-D] [RACKUP_FILE]
.SH DESCRIPTION
.PP
A rackup(1)\-like command to launch Rack applications using Unicorn.
It is expected to be started in your application root (APP_ROOT),
but the "working_directory" directive may be used in the CONFIG_FILE.
.PP
While unicorn takes a myriad of command\-line options for
compatibility with ruby(1) and rackup(1), it is recommended to stick
to the few command\-line options specified in the SYNOPSIS and use
the CONFIG_FILE as much as possible.
.SH RACKUP FILE
.PP
This defaults to "config.ru" in APP_ROOT.  It should be the same
file used by rackup(1) and other Rack launchers, it uses the
\f[I]Rack::Builder\f[] DSL.
.PP
Embedded command\-line options are mostly parsed for compatibility
with rackup(1) but strongly discouraged.
.SH UNICORN OPTIONS
.TP
.B \-c, \-\-config\-file CONFIG_FILE
Path to the Unicorn\-specific config file.  The config file is
implemented as a Ruby DSL, so Ruby code may executed.
See the RDoc/ri for the \f[I]Unicorn::Configurator\f[] class for the full
list of directives available from the DSL.
Using an absolute path for for CONFIG_FILE is recommended as it
makes multiple instances of Unicorn easily distinguishable when
viewing ps(1) output.
.RS
.RE
.TP
.B \-D, \-\-daemonize
Run daemonized in the background.  The process is detached from
the controlling terminal and stdin is redirected to "/dev/null".
Unlike many common UNIX daemons, we do not chdir to "/"
upon daemonization to allow more control over the startup/upgrade
process.
Unless specified in the CONFIG_FILE, stderr and stdout will
also be redirected to "/dev/null".
.RS
.RE
.TP
.B \-E, \-\-env RACK_ENV
Run under the given RACK_ENV.  See the RACK ENVIRONMENT section
for more details.
.RS
.RE
.TP
.B \-l, \-\-listen ADDRESS
Listens on a given ADDRESS.  ADDRESS may be in the form of
HOST:PORT or PATH, HOST:PORT is taken to mean a TCP socket
and PATH is meant to be a path to a UNIX domain socket.
Defaults to "0.0.0.0:8080" (all addresses on TCP port 8080)
For production deployments, specifying the "listen" directive in
CONFIG_FILE is recommended as it allows fine\-tuning of socket
options.
.RS
.RE
.TP
.B \-N, \-\-no\-default\-middleware
Disables loading middleware implied by RACK_ENV.  This bypasses the
configuration documented in the RACK ENVIRONMENT section, but still
allows RACK_ENV to be used for application/framework\-specific purposes.
.RS
.RE
.SH RACKUP COMPATIBILITY OPTIONS
.TP
.B \-o, \-\-host HOST
Listen on a TCP socket belonging to HOST, default is
"0.0.0.0" (all addresses).
If specified multiple times on the command\-line, only the
last\-specified value takes effect.
This option only exists for compatibility with the rackup(1) command,
use of "\-l"/"\-\-listen" switch is recommended instead.
.RS
.RE
.TP
.B \-p, \-\-port PORT
Listen on the specified TCP PORT, default is 8080.
If specified multiple times on the command\-line, only the last\-specified
value takes effect.
This option only exists for compatibility with the rackup(1) command,
use of "\-l"/"\-\-listen" switch is recommended instead.
.RS
.RE
.TP
.B \-s, \-\-server SERVER
No\-op, this exists only for compatibility with rackup(1).
.RS
.RE
.SH RUBY OPTIONS
.TP
.B \-e, \-\-eval LINE
Evaluate a LINE of Ruby code.  This evaluation happens
immediately as the command\-line is being parsed.
.RS
.RE
.TP
.B \-d, \-\-debug
Turn on debug mode, the $DEBUG variable is set to true.
.RS
.RE
.TP
.B \-w, \-\-warn
Turn on verbose warnings, the $VERBOSE variable is set to true.
.RS
.RE
.TP
.B \-I, \-\-include PATH
specify $LOAD_PATH.  PATH will be prepended to $LOAD_PATH.
The \[aq]:\[aq] character may be used to delimit multiple directories.
This directive may be used more than once.  Modifications to
$LOAD_PATH take place immediately and in the order they were
specified on the command\-line.
.RS
.RE
.TP
.B \-r, \-\-require LIBRARY
require a specified LIBRARY before executing the application.  The
"require" statement will be executed immediately and in the order
they were specified on the command\-line.
.RS
.RE
.SH SIGNALS
.PP
The following UNIX signals may be sent to the master process:
.IP \[bu] 2
HUP \- reload config file, app, and gracefully restart all workers
.IP \[bu] 2
INT/TERM \- quick shutdown, kills all workers immediately
.IP \[bu] 2
QUIT \- graceful shutdown, waits for workers to finish their
current request before finishing.
.IP \[bu] 2
USR1 \- reopen all logs owned by the master and all workers
See Unicorn::Util.reopen_logs for what is considered a log.
.IP \[bu] 2
USR2 \- reexecute the running binary.  A separate QUIT
should be sent to the original process once the child is verified to
be up and running.
.IP \[bu] 2
WINCH \- gracefully stops workers but keep the master running.
This will only work for daemonized processes.
.IP \[bu] 2
TTIN \- increment the number of worker processes by one
.IP \[bu] 2
TTOU \- decrement the number of worker processes by one
.PP
See the SIGNALS (https://yhbt.net/unicorn/SIGNALS.html) document for
full description of all signals used by Unicorn.
.SH RACK ENVIRONMENT
.PP
Accepted values of RACK_ENV and the middleware they automatically load
(outside of RACKUP_FILE) are exactly as those in rackup(1):
.IP \[bu] 2
development \- loads Rack::CommonLogger, Rack::ShowExceptions, and
              Rack::Lint middleware
.IP \[bu] 2
deployment \- loads Rack::CommonLogger middleware
.IP \[bu] 2
none \- loads no middleware at all, relying entirely on RACKUP_FILE
.PP
All unrecognized values for RACK_ENV are assumed to be
"none".  Production deployments are strongly encouraged to use
"deployment" or "none" for maximum performance.
.PP
As of Unicorn 0.94.0, RACK_ENV is exported as a process\-wide environment
variable as well.  While not current a part of the Rack specification as
of Rack 1.0.1, this has become a de facto standard in the Rack world.
.PP
Note the Rack::ContentLength middleware is also
loaded by "deployment" and "development", but no other values of
RACK_ENV.  If needed, they must be individually specified in the
RACKUP_FILE, some frameworks do not require them.
.SH ENVIRONMENT VARIABLES
.PP
The RACK_ENV variable is set by the aforementioned \-E switch.
All application or library\-specific environment variables (e.g. TMPDIR)
may always be set in the Unicorn CONFIG_FILE in addition to the spawning
shell.  When transparently upgrading Unicorn, all environment variables
set in the old master process are inherited by the new master process.
Unicorn only uses (and will overwrite) the UNICORN_FD environment
variable internally when doing transparent upgrades.
.PP
UNICORN_FD is a comma\-delimited list of one or more file descriptors
used to implement USR2 upgrades.  Init systems may bind listen sockets
itself and spawn unicorn with UNICORN_FD set to the file descriptor
numbers of the listen socket(s).
.PP
As of unicorn 5.0, LISTEN_PID and LISTEN_FDS are used for socket
activation as documented in the sd_listen_fds(3) manpage.  Users
relying on this feature do not need to specify a listen socket in
the unicorn config file.
.SH SEE ALSO
.IP \[bu] 2
\f[I]Rack::Builder\f[] ri/RDoc
.IP \[bu] 2
\f[I]Unicorn::Configurator\f[] ri/RDoc
.UR https://yhbt.net/unicorn/Unicorn/Configurator.html
.UE
.IP \[bu] 2
unicorn RDoc
.UR https://yhbt.net/unicorn/
.UE
.IP \[bu] 2
Rack RDoc
.UR https://www.rubydoc.info/github/rack/rack/
.UE
.IP \[bu] 2
Rackup HowTo
.UR https://github.com/rack/rack/wiki/(tutorial)-rackup-howto
.UE
.SH AUTHORS
The Unicorn Community <unicorn-public@yhbt.net>.


CHANGELOG
v0.10.3r - small fixes, depend on rev 0.3.0+
v0.10.2r - monkey-patch Rev for now (experimental)
v0.10.1r - --rainbows/-R command-line option (experimental)
v0.10.0r - Rainbows! (experimental, not going to Rubyforge)
v0.90.0 - switch chunking+trailer handling to Ragel, v0.8.4 fixes
v0.9.2 - Ruby 1.9.2 preview1 compatibility
v0.9.1 - FD_CLOEXEC portability fix (v0.8.2 port)
v0.9.0 - bodies: "Transfer-Encoding: chunked", rewindable streaming
v0.8.4 - pass through unknown HTTP status codes
v0.8.3 - Ruby 1.9.2 preview1 compatibility
v0.8.2 - socket handling bugfixes and usability tweaks
v0.8.1 - safer timeout handling, more consistent reload behavior
v0.8.0 - enforce Rack dependency, minor performance improvements and fixes
v0.7.1 - minor fixes, cleanups and documentation improvements
v0.7.0 - rack.version is 1.0
v0.6.0 - cleanups + optimizations, signals to {in,de}crement processes
v0.5.4 - fix data corruption with some small uploads (not curl)
v0.5.3 - fix 100% CPU usage when idle, small cleanups
v0.5.2 - force Status: header for compat, small cleanups
v0.5.1 - exit correctly on INT/TERM, QUIT is still recommended, however
v0.5.0 - {after,before}_fork API change, small tweaks/fixes
v0.4.2 - fix Rails ARStore, FD leak prevention, descriptive proctitles
v0.4.1 - Rails support, per-listener backlog and {snd,rcv}buf
v0.2.3 - Unlink Tempfiles after use (they were closed, just not unlinked)
v0.2.2 - small bug fixes, fix Rack multi-value headers (Set-Cookie:)
v0.2.1 - Fix broken Manifest that cause unicorn_rails to not be bundled
v0.2.0 - unicorn_rails launcher script.
v0.1.0 - Unicorn - UNIX-only fork of Mongrel free of threading


lib/unicorn.rb
# -*- encoding: binary -*-
require 'etc'
require 'stringio'
require 'kgio'
require 'raindrops'
require 'io/wait'

begin
  require 'rack'
rescue LoadError
  warn 'rack not available, functionality reduced'
end

# :stopdoc:
# Unicorn module containing all of the classes (include C extensions) for
# running a Unicorn web server.  It contains a minimalist HTTP server with just
# enough functionality to service web application requests fast as possible.
# :startdoc:

# unicorn exposes very little of an user-visible API and most of its
# internals are subject to change.  unicorn is designed to host Rack
# applications, so applications should be written against the Rack SPEC
# and not unicorn internals.
module Unicorn

  # Raised inside TeeInput when a client closes the socket inside the
  # application dispatch.  This is always raised with an empty backtrace
  # since there is nothing in the application stack that is responsible
  # for client shutdowns/disconnects.  This exception is visible to Rack
  # applications unless PrereadInput middleware is loaded.  This
  # is a subclass of the standard EOFError class and applications should
  # not rescue it explicitly, but rescue EOFError instead.
  ClientShutdown = Class.new(EOFError)

  # :stopdoc:

  # This returns a lambda to pass in as the app, this does not "build" the
  # app (which we defer based on the outcome of "preload_app" in the
  # Unicorn config).  The returned lambda will be called when it is
  # time to build the app.
  def self.builder(ru, op)
    # allow Configurator to parse cli switches embedded in the ru file
    op = Unicorn::Configurator::RACKUP.merge!(:file => ru, :optparse => op)
    if ru =~ /\.ru$/ && !defined?(Rack::Builder)
      abort "rack and Rack::Builder must be available for processing #{ru}"
    end

    # always called after config file parsing, may be called after forking
    lambda do |_, server|
      inner_app = case ru
      when /\.ru$/
        raw = File.read(ru)
        raw.sub!(/^__END__\n.*/, '')
        eval("Rack::Builder.new {(\n#{raw}\n)}.to_app", TOPLEVEL_BINDING, ru)
      else
        require ru
        Object.const_get(File.basename(ru, '.rb').capitalize)
      end

      if $DEBUG
        require 'pp'
        pp({ :inner_app => inner_app })
      end

      return inner_app unless server.default_middleware

      middleware = { # order matters
        ContentLength: nil,
        CommonLogger: [ $stderr ],
        ShowExceptions: nil,
        Lint: nil,
        TempfileReaper: nil,
      }

      # return value, matches rackup defaults based on env
      # Unicorn does not support persistent connections, but Rainbows!
      # does.  Users accustomed to the Rack::Server default
      # middlewares will need ContentLength middleware.
      case ENV["RACK_ENV"]
      when "development"
      when "deployment"
        middleware.delete(:ShowExceptions)
        middleware.delete(:Lint)
      else
        return inner_app
      end
      Rack::Builder.new do
        middleware.each do |m, args|
          use(Rack.const_get(m), *args) if Rack.const_defined?(m)
        end
        run inner_app
      end.to_app
    end
  end

  # returns an array of strings representing TCP listen socket addresses
  # and Unix domain socket paths.  This is useful for use with
  # Raindrops::Middleware under Linux: https://yhbt.net/raindrops/
  def self.listener_names
    Unicorn::HttpServer::LISTENERS.map do |io|
      Unicorn::SocketHelper.sock_name(io)
    end + Unicorn::HttpServer::NEW_LISTENERS
  end

  def self.log_error(logger, prefix, exc)
    message = exc.message
    message = message.dump if /[[:cntrl:]]/ =~ message
    logger.error "#{prefix}: #{message} (#{exc.class})"
    exc.backtrace.each { |line| logger.error(line) }
  end

  F_SETPIPE_SZ = 1031 if RUBY_PLATFORM =~ /linux/

  def self.pipe # :nodoc:
    Kgio::Pipe.new.each do |io|
      # shrink pipes to minimize impact on /proc/sys/fs/pipe-user-pages-soft
      # limits.
      if defined?(F_SETPIPE_SZ)
        begin
          io.fcntl(F_SETPIPE_SZ, Raindrops::PAGE_SIZE)
        rescue Errno::EINVAL
          # old kernel
        rescue Errno::EPERM
          # resizes fail if Linux is close to the pipe limit for the user
          # or if the user does not have permissions to resize
        end
      end
    end
  end
  # :startdoc:
end
# :enddoc:

%w(const socket_helper stream_input tee_input http_request configurator
   tmpio util http_response worker http_server).each do |s|
  require_relative "unicorn/#{s}"
end
